protocol/x/clob/memclob/memclob.go

```diff
@@ -1953,60 +1800,168 @@ func (m *MemClobPriceTimePriority) mustRemoveOrder(
+		offchainUpdates := types.NewOffchainUpdates()
 		if message, success := off_chain_updates.CreateOrderUpdateMessage(
 			ctx,
-			makerOrder.OrderId,
-			newTotalFilledAmount,
+			orderId,
+			filledQuantums,
 		); success {
-			offchainUpdates.AddUpdateMessage(makerOrder.OrderId, message)
+			offchainUpdates.AddUpdateMessage(orderId, message)
+			if m.generateOrderbookUpdates {
+				m.clobKeeper.SendOrderbookUpdates(ctx, offchainUpdates)
+			}
+			if m.generateOffchainUpdates {
+				m.clobKeeper.SendOffchainMessages(ctx, nil, offchainUpdates)
+			}
 		}
 	}
+}
+
+// mustUpdateOrderbookStateWithMatches updates the orderbook with matches.
+// If a maker order is fully filled, it removes it from the orderbook.
+func (m *MemClobPriceTimePriority) mustUpdateOrderbookStateWithMatches(
+	ctx sdk.Context,
+	match *types.ClobMatch,
+	takerOrder types.MatchableOrder,
+	makerFills []types.MakerFillWithOrder,
+) {
+	if m.generateOrderbookUpdates {
+		makerOrders := lib.MapSlice(
+			makerFills,
+			func(mfwo types.MakerFillWithOrder) types.Order {
+				return mfwo.Order
+			},
+		)
+		orderbookMatchFill := m.GenerateStreamOrderbookFill(ctx, *match, takerOrder, makerOrders)
+		m.clobKeeper.SendOrderbookFillUpdates(ctx, []types.StreamOrderbookFill{orderbookMatchFill})
+	}
 
-	// If the order is fully filled, remove it from the orderbook.
-	// Note we shouldn't remove Short-Term order hashes from `ShortTermOrderTxBytes` here since
-	// the order was matched.
-	if newTotalFilledAmount == makerOrderBaseQuantums {
-		makerOrderId := makerOrder.OrderId
-		m.mustRemoveOrder(ctx, makerOrderId)
+	for _, makerFill := range makerFills {
+		makerOrderBaseQuantums := makerFill.Order.GetBaseQuantums()
+		newTotalFilledAmount := m.GetOrderFilledAmount(ctx, makerFill.Order.OrderId)
 
+		// If the filled amount of the maker order is greater than the order size, panic to avoid silent failure.
+		if newTotalFilledAmount > makerOrderBaseQuantums {
+			panic("Total filled size of maker order greater than the order size")
+		}
+
+		offchainUpdates := types.NewOffchainUpdates()
 		if m.generateOffchainUpdates {
-			if message, success := off_chain_updates.CreateOrderRemoveMessageWithReason(
+			if message, success := off_chain_updates.CreateOrderUpdateMessage(
+				ctx,
+				makerFill.Order.OrderId,
+				newTotalFilledAmount,
+			); success {
+				offchainUpdates.AddUpdateMessage(makerFill.Order.OrderId, message)
+			}
+			m.clobKeeper.SendOffchainMessages(ctx, nil, offchainUpdates)
+		}
+
+		// If the order is fully filled, remove it from the orderbook.
+		// Note we shouldn't remove Short-Term order hashes from `ShortTermOrderTxBytes` here since
+		// the order was matched.
+		if newTotalFilledAmount == makerOrderBaseQuantums {
+			makerOrderId := makerFill.Order.OrderId
+			m.mustRemoveOrder(
 				ctx,
 				makerOrderId,
 				indexersharedtypes.OrderRemovalReason_ORDER_REMOVAL_REASON_FULLY_FILLED,
 				ocutypes.OrderRemoveV1_ORDER_REMOVAL_STATUS_BEST_EFFORT_CANCELED,
```

Comment:
Same as above, the order shouldn't be removed for being fully-filled due to in-memory matching.

Reply:
Talked offline and resolved, see above.